pkg/kubelet/status/status_manager.go

```diff
@@ -710,60 +687,84 @@ func (m *manager) RemoveOrphanedStatuses(podUIDs map[types.UID]bool) {
 	}
 }
 
-// syncBatch syncs pods statuses with the apiserver.
-func (m *manager) syncBatch() {
-	var updatedStatuses []podStatusSyncRequest
+// syncBatch syncs pods statuses with the apiserver. Returns the number of syncs
+// attempted for testing.
+func (m *manager) syncBatch(all bool) int {
+	type podSync struct {
+		podUID    types.UID
+		statusUID kubetypes.MirrorPodUID
+		status    versionedPodStatus
+	}
+
+	var updatedStatuses []podSync
 	podToMirror, mirrorToPod := m.podManager.GetUIDTranslations()
 	func() { // Critical section
 		m.podStatusesLock.RLock()
 		defer m.podStatusesLock.RUnlock()
 
 		// Clean up orphaned versions.
-		for uid := range m.apiStatusVersions {
-			_, hasPod := m.podStatuses[types.UID(uid)]
-			_, hasMirror := mirrorToPod[uid]
-			if !hasPod && !hasMirror {
-				delete(m.apiStatusVersions, uid)
+		if all {
+			for uid := range m.apiStatusVersions {
+				_, hasPod := m.podStatuses[types.UID(uid)]
+				_, hasMirror := mirrorToPod[uid]
+				if !hasPod && !hasMirror {
+					delete(m.apiStatusVersions, uid)
+				}
 			}
 		}
 
+		// Decide which pods need status updates.
 		for uid, status := range m.podStatuses {
-			syncedUID := kubetypes.MirrorPodUID(uid)
+			// translate the pod UID (source) to the status UID (API pod) -
+			// static pods are identified in source by pod UID but tracked in the
+			// API via the uid of the mirror pod
+			uidOfStatus := kubetypes.MirrorPodUID(uid)
 			if mirrorUID, ok := podToMirror[kubetypes.ResolvedPodUID(uid)]; ok {
 				if mirrorUID == "" {
 					klog.V(5).InfoS("Static pod does not have a corresponding mirror pod; skipping",
 						"podUID", uid,
 						"pod", klog.KRef(status.podNamespace, status.podName))
 					continue
 				}
-				syncedUID = mirrorUID
+				uidOfStatus = mirrorUID
+			}
+
+			// if a new status update has been delivered, trigger an update, otherwise the
+			// pod can wait for the next bulk check (which performs reconciliation as well)
+			if !all {
+				if m.apiStatusVersions[uidOfStatus] >= status.version {
```

Comment:
Is it necessary here to check if the status also exists explicitly? I suppose if `status.version` starts out at 1 it would work as well as we could rely on the default value from, but it's not immediately clear if that will work. i.e. something like:

```
lastStatusVersion, ok := m.apiStatusVersions[uidOfStatus]
if ok && lastStatusVersion >= status.version {
    continue
}
...
```

Reply:
also, looking at this again, I'm curious is there any reason we can't reuse `needsUpdate` check here? i.e. similar to the previous logic of

```
if !all {
    if needsUpdate(uid, status) {
        updatedStatuses = append(updatedStatuses, podStatusSyncRequest{uid, status})
    }
}
```

Then, the only difference between `!all` and `all` would be that `all` performs `needsUpdate` and `needsReconcile` while `!all` just performs `needsUpdate`

Reply:
There are three parts of needsUpdate

1. Check whether the statusVersion is newer or absent (return true if so)
2. Check whether the pod is in podManager (return false if not)
3. Check whether the pod can be deleted (return true if so)

2 is redundant - we just fetched pod from podManager (above) and so the race window is as short as if we checked podManager inside syncPod (RTT for a single GET -> PATCH is usually 40-50ms, if we processed 5 pods maybe we would race, but we mostly don't care.  Previously, we *didn't* check podManager, and the status channel could be backed up.  But it's not really an issue to execute a GET or PATCH for already deleted pod, and we want to get this logic simpler before we start doing parallelism (which is a more effective way to remove the races).

3 is not appropriate in this case - we're not reconciling.

I would probably say that needsUpdate is going to be removed for a switch statement in a future revision, and when we parallelize or SSA the pod we'll also see a drop in latency.  And remember - this only applies to force deleted pods, which should be vanishingly rare in production, so I'd rather have a simpler logic path that we can adapt more easily to other code.

I could remove needsUpdate in this PR but better to that after master reopens (I think).

Reply:
Got it, thanks for the explanation, makes sense then to avoid callign `needsUpdate` since it sounds redundant. Do we still need to explicitly check if the status exists in `apiStatusVersions`? It seems like we should only `continue` if the status both exists in the map and the version is API status version is `>=` the `status.version`?

Also, I didn't completely follow your comment regarding "this only applies to force deleted pods". I thought this logic was being applied for all statuses updates because we are looping over `m.podStatuses`? Or am I misunderstading somewhere why this logic is only applied for force deleted pods?

Reply:
> Do we still need to explicitly check if the status exists in apiStatusVersions?

If apiStatusVersions doesn't have a UID, its value in the map is 0, which will be `< status.Version`, triggering an update.

A pod can't be removed from podManager unless the kubelet completes deletion (the `syncPod` method at the end) OR someone force deletes the pod, which is what case 2 is checking. For static pods we don't really care about status updates anyway (mirror pods can be pretty laggy).  Regular pods shouldn't be force deleted that much.

Reply:
Thanks for explanation and double checking that `version > 0`, all makes sense.